lib/srv/desktop/rdp/rdpclient/src/vchan.rs

```diff
@@ -13,10 +13,76 @@
 // limitations under the License.
 
 use crate::errors::invalid_data_error;
-use crate::Payload;
+use crate::RawPayload;
 use bitflags::bitflags;
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use rdp::core::mcs;
+use rdp::core::tpkt;
 use rdp::model::error::*;
+use rdp::try_let;
+use std::io::{Cursor, Read, Write};
+
+/// Client is a general client for handling virtual channel payloads.
+/// It's read method can read an RDP message sent in multiple chunks
+/// (or a single chunk) over a virtual channel.
+/// See https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/343e4888-4c48-4054-b0e3-4e0762d1993c
+/// for more information about chunks.
+pub struct Client<C: VchanClient> {
```

Comment:
I don't understand why we need a generic type param here.

vchan should just piece out chunks and hand them off to something else, right? Can we do that without this extra complexity of `Client<Client<...>>`

I'm thinking of some flow like this:

``` 
let v = vchan::Client::new()

let message: Option<Payload> = v.read(payload)
// returns Some(Payload) if this is a full chunk, 
// otherwise returns None

if let Some(payload) = message {
  cliprdr.handle_message(payload)
}
```

Reply:
Yeah, I have an iteration that's like this, but then I got carried away with having generics at my disposal and turned it into this. I will revert to the other version and polish it off.